File: ts/src/core/execution/nodes/telegramListener.ts
/**
 * TelegramListenerNode – autonomous node that polls Telegram for new messages.
 * Mirrors Python src/core/execution/nodes/telegram_listener.py
 *
 * This is a CONTINUOUS node. The WorkflowRunner calls onTick() every ~100ms;
 * the node respects its own poll_interval to avoid hammering the Telegram API.
 * When new messages arrive they are queued and emitted one-per-tick so each
 * message gets its own full downstream graph execution.
 */
import { BaseNode, ExecutionContext, ExecutionMode } from "../nodeBase";
import { WorkflowData, NodeID } from "../../types";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("telegramListener");

const API_BASE = "https://api.telegram.org/bot";

interface ParsedMessage {
  message: string;
  user_id: string;
  username: string;
  first_name: string;
  chat_id: string;
  chat_type: string;
  chat_title: string;
  message_id: number;
  is_reply_to_bot: boolean;
  is_mention: boolean;
  timestamp: number;
  raw_update: Record<string, unknown>;
}

export class TelegramListenerNode extends BaseNode {
  // ── CONTINUOUS execution mode ──────────────────────────────────────
  static override executionMode = ExecutionMode.CONTINUOUS;

  // ── Instance state (persists across ticks) ─────────────────────────
  private _botToken = "";
  private _pollInterval: number; // seconds
  private _timeout: number; // seconds
  private _lastUpdateId: number | null = null;
  private _lastPollTime = 0;
  private _messageCount = 0;
  private _botInfo: Record<string, unknown> | null = null;
  private _pendingMessages: ParsedMessage[] = [];

  constructor(nodeId: string, nodeData: import("../../types").NodeData) {
    super(nodeId, nodeData);

    const meta = this.metadata;
    this._botToken =
      (this.resolveEnvVar(meta.bot_token) as string) ||
      process.env.TELEGRAM_DEV_AGENT_BOT_TOKEN ||
      process.env.TELEGRAM_BOT_TOKEN ||
      "";
    this._pollInterval = Number(meta.poll_interval ?? 2);
    this._timeout = Number(meta.timeout ?? 30);

    logger.debug(
      `[TelegramListener ${nodeId}] Initialized: poll_interval=${this._pollInterval}s, timeout=${this._timeout}s`
    );
  }

  // ── initialize() — called once when runner starts the workflow ──────
  override async initialize(
    _workflow: WorkflowData,
    _allNodes: Map<NodeID, BaseNode>
  ): Promise<void> {
    // Fetch bot info so onTick can parse mentions / reply detection
    await this._getBotInfo();

    if (!this._botToken) {
      logger.warn("[TelegramListener] No bot_token — node will not poll");
    } else if (this._botInfo) {
      logger.info(
        `[TelegramListener] Ready to poll as @${(this._botInfo as any).username}`
      );
    }

    // Skip all pending updates so the bot only processes NEW messages
    await this._skipOldUpdates();

    // Seed timing so the first poll happens after poll_interval
    this._lastPollTime = Date.now() / 1000;
  }

  // ── execute() — returns current state (no side effects) ────────────
  async execute(_context: ExecutionContext): Promise<Record<string, unknown>> {
    return {
      trigger: false,
      message: "",
      user_id: "",
      username: "",
      chat_id: "",
      message_id: 0,
      is_reply_to_bot: false,
      is_mention: false,
      is_dm: false,
      raw_update: null,
    };
  }

  // ── onTick() — called every ~100ms by WorkflowRunner ──────────────
  async onTick(_context: ExecutionContext): Promise<Record<string, unknown> | null> {
    if (!this._botToken) return null;

    // 1. If we have pending messages from a previous poll, emit one
    if (this._pendingMessages.length) {
      return this._emitNextMessage();
    }

    // 2. Check if poll interval has elapsed
    const now = Date.now() / 1000;
    if (now - this._lastPollTime < this._pollInterval) return null;
    this._lastPollTime = now;

    // 3. Poll Telegram
    const updates = await this._getUpdates();
    if (!updates.length) return null;

    // 4. Parse all updates and queue them
    for (const update of updates) {
      const parsed = this._parseUpdate(update);
      if (parsed && parsed.message) {
        this._pendingMessages.push(parsed);
      }
    }

    if (this._pendingMessages.length) {
      logger.debug(
        `[TelegramListener ${this.nodeId}] Queued ${this._pendingMessages.length} messages for processing`
      );
      return this._emitNextMessage();
    }

    return null;
  }

  // ── Private helpers ────────────────────────────────────────────────

  /**
   * Fast-forward past all pending Telegram updates so the bot starts
   * fresh.  Calls getUpdates with offset=-1 to grab only the very
   * latest update, records its update_id, and discards the message.
   */
  private async _skipOldUpdates(): Promise<void> {
    if (!this._botToken) return;

    try {
      const params = new URLSearchParams({
        offset: "-1",
        limit: "1",
        timeout: "0",
      });

      const url = `${API_BASE}${this._botToken}/getUpdates?${params}`;
      const res = await fetch(url, { signal: AbortSignal.timeout(10_000) });
      const json = (await res.json()) as Record<string, unknown>;

      if (json.ok) {
        const updates = (json.result as Record<string, unknown>[]) ?? [];
        if (updates.length) {
          this._lastUpdateId = updates[updates.length - 1].update_id as number;
          logger.info(
            `[TelegramListener] Skipped old updates, starting after update_id=${this._lastUpdateId}`
          );
        } else {
          logger.info("[TelegramListener] No pending updates, starting fresh");
        }
      } else {
        const desc = (json.description as string) || JSON.stringify(json).slice(0, 300);
        logger.error(
          `[TelegramListener] Failed to skip old updates — API returned ok=false ` +
          `(HTTP ${res.status} ${res.statusText}): ${desc}`
        );
      }
    } catch (err) {
      logger.error(
        `[TelegramListener] Failed to skip old updates: ${err instanceof Error ? err.message : err}`
      );
    }
  }

  private _emitNextMessage(): Record<string, unknown> | null {
    if (!this._pendingMessages.length) return null;
    const parsed = this._pendingMessages.shift()!;
    this._messageCount++;

    logger.info(
      `[TelegramListener ${this.nodeId}] Message #${this._messageCount} ` +
        `from @${parsed.username || parsed.user_id} ` +
        `in ${parsed.chat_type} ${parsed.chat_id}: ` +
        `${parsed.message.slice(0, 50)}...`
    );

    return {
      trigger: true,
      message: parsed.message,
      user_id: parsed.user_id,
      username: parsed.username,
      chat_id: parsed.chat_id,
      message_id: parsed.message_id,
      is_reply_to_bot: parsed.is_reply_to_bot,
      is_mention: parsed.is_mention,
      is_dm: parsed.chat_type === "private",
      raw_update: parsed.raw_update,
    };
  }

  private async _getBotInfo(): Promise<Record<string, unknown> | null> {
    if (this._botInfo) return this._botInfo;
    if (!this._botToken) return null;

    try {
      const res = await fetch(`${API_BASE}${this._botToken}/getMe`, {
        signal: AbortSignal.timeout(10_000),
      });
      const json = (await res.json()) as Record<string, unknown>;
      if (json.ok) {
        this._botInfo = json.result as Record<string, unknown>;
        logger.info(
          `[TelegramListener] Bot info: @${(this._botInfo as any).username}`
        );
        return this._botInfo;
      }
    } catch (err) {
      logger.error(
        `[TelegramListener] Failed to get bot info: ${err instanceof Error ? err.message : err}`
      );
    }
    return null;
  }

  private async _getUpdates(): Promise<Record<string, unknown>[]> {
    if (!this._botToken) return [];

    try {
      const params = new URLSearchParams({
        timeout: String(this._timeout),
        allowed_updates: JSON.stringify(["message"]),
        limit: "10",
      });
      if (this._lastUpdateId !== null) {
        params.set("offset", String(this._lastUpdateId + 1));
      }

      const res = await fetch(
        `${API_BASE}${this._botToken}/getUpdates?${params}`,
        { signal: AbortSignal.timeout((this._timeout + 5) * 1000) }
      );
      const json = (await res.json()) as Record<string, unknown>;

      if (!json.ok) {
        logger.error(
          `[TelegramListener] API error: ${JSON.stringify(json).slice(0, 300)}`
        );
        return [];
      }

      const updates = (json.result as Record<string, unknown>[]) ?? [];
      if (updates.length) {
        this._lastUpdateId = Math.max(
          ...updates.map((u) => u.update_id as number)
        );
        logger.debug(
          `[TelegramListener] Got ${updates.length} updates, last_id=${this._lastUpdateId}`
        );
      }
      return updates;
    } catch (err) {
      if ((err as any)?.name === "TimeoutError") {
        logger.debug("[TelegramListener] Poll timeout (normal)");
      } else {
        logger.error(
          `[TelegramListener] Request error: ${err instanceof Error ? err.message : err}`
        );
      }
      return [];
    }
  }

  private _parseUpdate(update: Record<string, unknown>): ParsedMessage | null {
    const message = update.message as Record<string, unknown> | undefined;
    if (!message) return null;

    const text =
      (message.text as string) || (message.caption as string) || "";

    const fromUser = (message.from as Record<string, unknown>) ?? {};
    const userId = String(fromUser.id ?? "");
    const username = (fromUser.username as string) ?? "";
    const firstName = (fromUser.first_name as string) ?? "";

    const chat = (message.chat as Record<string, unknown>) ?? {};
    const chatId = String(chat.id ?? "");
    const chatType = (chat.type as string) ?? "";
    const chatTitle = (chat.title as string) ?? "";

    // Check if reply to bot
    const replyTo = (message.reply_to_message as Record<string, unknown>) ?? {};
    const replyFrom = (replyTo.from as Record<string, unknown>) ?? {};
    const isReplyToBot =
      !!this._botInfo && replyFrom.id === (this._botInfo as any).id;

    // Check if bot is @mentioned
    const botUsername = (this._botInfo as any)?.username ?? "";
    const isMention =
      botUsername ? text.toLowerCase().includes(`@${botUsername}`.toLowerCase()) : false;

    return {
      message: text,
      user_id: userId,
      username,
      first_name: firstName,
      chat_id: chatId,
      chat_type: chatType,
      chat_title: chatTitle,
      message_id: (message.message_id as number) ?? 0,
      is_reply_to_bot: isReplyToBot,
      is_mention: isMention,
      timestamp: (message.date as number) ?? 0,
      raw_update: update,
    };
  }
}

File: src/core/execution/nodes/telegram_listener.py
"""
Telegram Listener Node
Autonomous node that polls Telegram for new messages
"""
import time
from typing import Dict, Any, Optional, List
import requests

from ..node_base import BaseNode, ExecutionContext, ExecutionMode
from ....utils.logger import get_logger

logger = get_logger(__name__)


class TelegramListenerNode(BaseNode):
    """
    Autonomous node that listens for Telegram messages via long polling.
    
    This node runs continuously and polls the Telegram Bot API for new messages.
    When messages are received, it outputs them for processing by downstream nodes.
    
    Properties (from metadata):
        bot_token: Telegram bot API token (required)
        poll_interval: Seconds between polls (default: 2)
        timeout: Long polling timeout in seconds (default: 30)
        
    Outputs:
        trigger: Boolean - True when new message(s) received
        message: String - The latest message text
        user_id: String - Sender's Telegram user ID (for memory layer)
        username: String - Sender's @username (may be empty)
        chat_id: String - The chat/group ID
        message_id: Number - Telegram message ID
        is_reply_to_bot: Boolean - True if message is a reply to bot's message
        raw_update: Object - Full Telegram update object
    """
    
    # This is a CONTINUOUS node - runs on every tick
    execution_mode = ExecutionMode.CONTINUOUS
    
    # Telegram API base URL
    API_BASE = "https://api.telegram.org/bot"
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        super().__init__(node_id, node_data)
        
        # Get configuration from metadata
        self._bot_token = self.metadata.get('bot_token', '')
        self._poll_interval = float(self.metadata.get('poll_interval', 2.0))
        self._timeout = int(self.metadata.get('timeout', 30))
        
        # Resolve environment variables in bot_token
        self._bot_token = self._resolve_env_var(self._bot_token)
        
        # State
        self._last_update_id: Optional[int] = None
        self._last_poll_time: float = 0.0
        self._message_count: int = 0
        self._bot_info: Optional[Dict] = None
        self._pending_messages: List[Dict] = []  # Queue for messages from multi-message polls
        
        logger.debug(
            f"[TelegramListener {node_id}] Initialized: "
            f"poll_interval={self._poll_interval}s, "
            f"timeout={self._timeout}s"
        )
    
    def _resolve_env_var(self, value: str) -> str:
        """Resolve environment variable template like {{process.env.VAR_NAME}}"""
        import os
        if isinstance(value, str) and value.startswith('{{') and value.endswith('}}'):
            var_name = value[2:-2].strip()
            if var_name.startswith('process.env.'):
                env_var = var_name.replace('process.env.', '')
                resolved = os.getenv(env_var, '')
                if not resolved:
                    logger.warning(f"[TelegramListener] Env var {env_var} not found")
                return resolved
        return value
    
    def _get_bot_info(self) -> Optional[Dict]:
        """Get bot info (username, id) from Telegram API"""
        if self._bot_info:
            return self._bot_info
            
        if not self._bot_token:
            return None
            
        try:
            url = f"{self.API_BASE}{self._bot_token}/getMe"
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            result = response.json()
            
            if result.get('ok'):
                self._bot_info = result.get('result', {})
                logger.info(f"[TelegramListener] Bot info: @{self._bot_info.get('username')}")
                return self._bot_info
        except Exception as e:
            logger.error(f"[TelegramListener] Failed to get bot info: {e}")
        
        return None
    
    def _skip_old_updates(self) -> None:
        """
        Fast-forward past all pending updates so the bot starts fresh.
        
        Calls getUpdates with offset=-1 to fetch only the very latest update,
        then records its update_id so subsequent polls only see new messages.
        """
        if not self._bot_token:
            return
        
        try:
            url = f"{self.API_BASE}{self._bot_token}/getUpdates"
            params = {
                'offset': -1,
                'limit': 1,
                'timeout': 0,
            }
            response = requests.get(url, params=params, timeout=10)
            response.raise_for_status()
            result = response.json()
            
            if result.get('ok'):
                updates = result.get('result', [])
                if updates:
                    self._last_update_id = updates[-1].get('update_id', 0)
                    logger.info(
                        f"[TelegramListener] Skipped old updates, starting after update_id={self._last_update_id}"
                    )
                else:
                    logger.info("[TelegramListener] No pending updates, starting fresh")
            else:
                desc = result.get('description') or str(result)[:300]
                logger.error(
                    f"[TelegramListener] Failed to skip old updates — API returned ok=false "
                    f"(HTTP {response.status_code}): {desc}"
                )
        except Exception as e:
            logger.error(f"[TelegramListener] Failed to skip old updates: {e}")
    
    def _get_updates(self) -> List[Dict]:
        """
        Poll Telegram API for new updates using long polling.
        
        Returns:
            List of update objects
        """
        if not self._bot_token:
            logger.warning("[TelegramListener] No bot_token configured")
            return []
        
        try:
            url = f"{self.API_BASE}{self._bot_token}/getUpdates"
            params = {
                'timeout': self._timeout,
                'allowed_updates': ['message'],  # Only get message updates
            }
            
            # If we have a last update ID, only get newer updates
            if self._last_update_id is not None:
                params['offset'] = self._last_update_id + 1
            
            response = requests.get(url, params=params, timeout=self._timeout + 5)
            response.raise_for_status()
            result = response.json()
            
            if result.get('ok'):
                updates = result.get('result', [])
                
                # Update last_update_id to avoid getting same updates again
                if updates:
                    self._last_update_id = max(u.get('update_id', 0) for u in updates)
                    logger.debug(f"[TelegramListener] Got {len(updates)} updates, last_id={self._last_update_id}")
                
                return updates
            else:
                logger.error(f"[TelegramListener] API error: {result.get('description')}")
                return []
                
        except requests.exceptions.Timeout:
            # Timeout is normal for long polling
            logger.debug("[TelegramListener] Poll timeout (normal)")
            return []
        except requests.exceptions.RequestException as e:
            logger.error(f"[TelegramListener] Request error: {e}")
            return []
        except Exception as e:
            logger.error(f"[TelegramListener] Unexpected error: {e}")
            return []
    
    def _parse_update(self, update: Dict) -> Optional[Dict[str, Any]]:
        """
        Parse a Telegram update into a standardized format.
        
        Args:
            update: Raw Telegram update object
            
        Returns:
            Parsed message data or None if not a valid message
        """
        message = update.get('message')
        if not message:
            return None
        
        # Extract message text (could be text, caption, etc.)
        text = message.get('text') or message.get('caption') or ''
        
        # Extract sender info
        from_user = message.get('from', {})
        user_id = str(from_user.get('id', ''))
        username = from_user.get('username', '')
        first_name = from_user.get('first_name', '')
        
        # Extract chat info
        chat = message.get('chat', {})
        chat_id = str(chat.get('id', ''))
        chat_type = chat.get('type', '')  # 'private', 'group', 'supergroup', 'channel'
        chat_title = chat.get('title', '')
        
        # Check if this is a reply to bot's message
        reply_to = message.get('reply_to_message', {})
        reply_to_from = reply_to.get('from', {})
        bot_info = self._get_bot_info()
        is_reply_to_bot = False
        if bot_info and reply_to_from.get('id') == bot_info.get('id'):
            is_reply_to_bot = True
        
        # Check if bot is mentioned in the message
        bot_username = bot_info.get('username', '') if bot_info else ''
        is_mention = f"@{bot_username}".lower() in text.lower() if bot_username else False
        
        return {
            'message': text,
            'user_id': user_id,
            'username': username,
            'first_name': first_name,
            'chat_id': chat_id,
            'chat_type': chat_type,
            'chat_title': chat_title,
            'message_id': message.get('message_id'),
            'is_reply_to_bot': is_reply_to_bot,
            'is_mention': is_mention,
            'timestamp': message.get('date'),
            'raw_update': update
        }
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """
        Initial execution - get bot info and prepare for polling.
        """
        # Get bot info on startup
        bot_info = self._get_bot_info()
        
        if not self._bot_token:
            logger.warning("[TelegramListener] No bot_token - node will not poll")
        elif bot_info:
            logger.info(f"[TelegramListener] Ready to poll as @{bot_info.get('username')}")
        
        # Skip all pending updates so the bot only processes NEW messages
        self._skip_old_updates()
        
        # Initialize polling time
        self._last_poll_time = time.time()
        
        return {
            'trigger': False,
            'message': '',
            'user_id': '',
            'username': '',
            'chat_id': '',
            'message_id': 0,
            'is_reply_to_bot': False,
            'is_mention': False,
            'raw_update': None
        }
    
    def on_tick(self, context: ExecutionContext) -> Optional[Dict[str, Any]]:
        """
        Called on each tick to poll for new messages.
        
        Returns:
            Output dict with message data if new message received, None otherwise
        """
        if not self._bot_token:
            return None
        
        # First, check if we have pending messages from a previous poll
        if self._pending_messages:
            return self._emit_next_message()
        
        # No pending messages - check if we should poll
        current_time = time.time()
        elapsed = current_time - self._last_poll_time
        
        # Check if poll interval has elapsed
        if elapsed < self._poll_interval:
            return None
        
        self._last_poll_time = current_time
        
        # Poll for updates
        updates = self._get_updates()
        
        if not updates:
            return None
        
        # Parse ALL updates and queue them
        for update in updates:
            parsed = self._parse_update(update)
            if parsed and parsed.get('message'):
                self._pending_messages.append(parsed)
        
        # Log how many messages were queued
        if self._pending_messages:
            logger.debug(f"[TelegramListener {self.node_id}] Queued {len(self._pending_messages)} messages for processing")
        
        # Emit the first message if any
        if self._pending_messages:
            return self._emit_next_message()
        
        return None
    
    def _emit_next_message(self) -> Optional[Dict[str, Any]]:
        """
        Pop and emit the next message from the pending queue.
        
        Returns:
            Output dict with message data, or None if queue is empty
        """
        if not self._pending_messages:
            return None
        
        parsed = self._pending_messages.pop(0)
        self._message_count += 1
        
        logger.info(
            f"[TelegramListener {self.node_id}] Message #{self._message_count} "
            f"from @{parsed.get('username') or parsed.get('user_id')} "
            f"in {parsed.get('chat_type')} {parsed.get('chat_id')}: "
            f"{parsed.get('message')[:50]}..."
        )
        
        return {
            'trigger': True,
            'message': parsed.get('message', ''),
            'user_id': parsed.get('user_id', ''),
            'username': parsed.get('username', ''),
            'chat_id': parsed.get('chat_id', ''),
            'message_id': parsed.get('message_id', 0),
            'is_reply_to_bot': parsed.get('is_reply_to_bot', False),
            'is_mention': parsed.get('is_mention', False),
            'raw_update': parsed.get('raw_update')
        }
    
    def get_status(self) -> Dict[str, Any]:
        """Get current listener status"""
        return {
            'bot_token_set': bool(self._bot_token),
            'bot_username': self._bot_info.get('username') if self._bot_info else None,
            'message_count': self._message_count,
            'last_update_id': self._last_update_id,
            'poll_interval': self._poll_interval,
            'timeout': self._timeout
        }
